Crash in QR scanner when camera no longer available

This can happen temporarily and needs to be caught / logged. Right now it brings down the app.

```
 android.hardware.camera2.CameraAccessException: CAMERA_DISCONNECTED (2): submitRequestList:137: Camera device no longer alive
        at android.hardware.camera2.CameraManager.throwAsPublicException(CameraManager.java:660)
        at android.hardware.camera2.impl.ICameraDeviceUserWrapper.submitRequestList(ICameraDeviceUserWrapper.java:86)
        at android.hardware.camera2.impl.CameraDeviceImpl.submitCaptureRequest(CameraDeviceImpl.java:899)
        at android.hardware.camera2.impl.CameraDeviceImpl.setRepeatingRequest(CameraDeviceImpl.java:938)
        at android.hardware.camera2.impl.CameraCaptureSessionImpl.setRepeatingRequest(CameraCaptureSessionImpl.java:243)
        at mozilla.components.feature.qr.QrFragment$createCameraPreviewSession$$inlined$let$lambda$1.onConfigured(QrFragment.kt:428)
        at java.lang.reflect.Method.invoke(Native Method)
        at android.hardware.camera2.dispatch.InvokeDispatcher.dispatch(InvokeDispatcher.java:39)
        at android.hardware.camera2.dispatch.HandlerDispatcher$1.run(HandlerDispatcher.java:65)
        at android.os.Handler.handleCallback(Handler.java:751)
        at android.os.Handler.dispatchMessage(Handler.java:95)
        at android.os.Looper.loop(Looper.java:154)
        at android.os.HandlerThread.run(HandlerThread.java:61)
     Caused by: android.os.ServiceSpecificException: submitRequestList:137: Camera device no longer alive
        at android.os.Parcel.readException(Parcel.java:1698)
        at android.os.Parcel.readException(Parcel.java:1637)
        at android.hardware.camera2.ICameraDeviceUser$Stub$Proxy.submitRequestList(ICameraDeviceUser.java:334)
        at android.hardware.camera2.impl.ICameraDeviceUserWrapper.submitRequestList(ICameraDeviceUserWrapper.java:84)
```

--- a/components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt
+++ b/components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt
@@ -23,6 +23,7 @@ import android.graphics.Matrix
 import android.graphics.Point
 import android.graphics.RectF
 import android.graphics.SurfaceTexture
+import android.hardware.camera2.CameraAccessException
 import android.hardware.camera2.CameraCaptureSession
 import android.hardware.camera2.CameraCharacteristics
 import android.hardware.camera2.CameraDevice
@@ -49,7 +50,7 @@ import com.google.zxing.NotFoundException
 import com.google.zxing.PlanarYUVLuminanceSource
 import com.google.zxing.common.HybridBinarizer
 import mozilla.components.support.base.android.view.AutoFitTextureView
-import mozilla.components.support.base.log.Log
+import mozilla.components.support.base.log.logger.Logger
 import java.io.Serializable
 import java.lang.IllegalStateException
 
@@ -68,8 +69,10 @@ import java.util.concurrent.TimeUnit
  * https://github.com/googlesamples/android-Camera2Basic
  * https://github.com/kismkof/camera2basic
  */
-@Suppress("TooManyFunctions")
+@Suppress("TooManyFunctions", "LargeClass")
 class QrFragment : Fragment() {
+    private val logger = Logger("mozac-qr")
+
     /**
      * [TextureView.SurfaceTextureListener] handles several lifecycle events on a [TextureView].
      */
@@ -94,7 +97,7 @@ class QrFragment : Fragment() {
     private var cameraId: String? = null
     private var captureSession: CameraCaptureSession? = null
     internal var cameraDevice: CameraDevice? = null
-    private var previewSize: Size? = null
+    internal var previewSize: Size? = null
 
     /**
      * Listener invoked when the QR scan completed successfully.
@@ -238,7 +241,7 @@ class QrFragment : Fragment() {
             backgroundThread = null
             backgroundHandler = null
         } catch (e: InterruptedException) {
-            Log.log(Log.Priority.DEBUG, "mozac-qr", e, "Interrupted while stopping background thread")
+            logger.debug("Interrupted while stopping background thread", e)
         }
     }
 
@@ -263,7 +266,7 @@ class QrFragment : Fragment() {
             val map = characteristics.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP) ?: continue
             val largest = Collections.max(map.getOutputSizes(ImageFormat.YUV_420_888).asList(), CompareSizesByArea())
             imageReader = ImageReader.newInstance(MAX_PREVIEW_WIDTH, MAX_PREVIEW_HEIGHT, ImageFormat.YUV_420_888, 2)
-                .apply { setOnImageAvailableListener(imageAvailableListener, backgroundHandler) }
+                    .apply { setOnImageAvailableListener(imageAvailableListener, backgroundHandler) }
 
             // Find out if we need to swap dimension to get the preview size relative to sensor coordinate.
             val displayRotation = activity?.windowManager?.defaultDisplay?.rotation
@@ -394,7 +397,8 @@ class QrFragment : Fragment() {
     /**
      * Creates a new [CameraCaptureSession] for camera preview.
      */
-    private fun createCameraPreviewSession() {
+    @Suppress("ComplexMethod")
+    internal fun createCameraPreviewSession() {
         val texture = textureView.surfaceTexture
 
         val size = previewSize as Size
@@ -403,15 +407,16 @@ class QrFragment : Fragment() {
 
         val surface = Surface(texture)
         val mImageSurface = imageReader?.surface
-        cameraDevice?.let {
-            previewRequestBuilder = it.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW).apply {
-                addTarget(mImageSurface as Surface)
-                addTarget(surface)
-            }
 
-            it.createCaptureSession(Arrays.asList(mImageSurface, surface),
-                object : CameraCaptureSession.StateCallback() {
+        try {
+            cameraDevice?.let {
+                previewRequestBuilder = it.createCaptureRequest(CameraDevice.TEMPLATE_PREVIEW).apply {
+                    addTarget(mImageSurface as Surface)
+                    addTarget(surface)
+                }
 
+                val captureCallback = object : CameraCaptureSession.CaptureCallback() {}
+                val stateCallback = object : CameraCaptureSession.StateCallback() {
                     override fun onConfigured(cameraCaptureSession: CameraCaptureSession) {
                         if (null == cameraDevice) return
 
@@ -419,19 +424,28 @@ class QrFragment : Fragment() {
                                 CaptureRequest.CONTROL_AF_MODE_CONTINUOUS_PICTURE)
 
                         previewRequest = previewRequestBuilder?.build()
-
                         captureSession = cameraCaptureSession
-                        cameraCaptureSession.setRepeatingRequest(previewRequest as CaptureRequest,
-                            object : CameraCaptureSession.CaptureCallback() {},
-                            backgroundHandler
-                        )
+
+                        try {
+                            cameraCaptureSession.setRepeatingRequest(
+                                    previewRequest as CaptureRequest,
+                                    captureCallback,
+                                    backgroundHandler
+                            )
+                        } catch (e: CameraAccessException) {
+                            logger.error("Failed to request capture", e)
+                        }
                     }
 
                     override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {
-                        Log.log(Log.Priority.ERROR, "mozac-qr", message = "Failed to configure CameraCaptureSession")
+                        logger.error("Failed to configure CameraCaptureSession")
                     }
-                }, null
-            )
+                }
+
+                it.createCaptureSession(Arrays.asList(mImageSurface, surface), stateCallback, null)
+            }
+        } catch (e: CameraAccessException) {
+                logger.error("Failed to create camera preview session", e)
         }
     }
 

--- a/components/feature/qr/src/test/java/mozilla/components/feature/qr/QrFragmentTest.kt
+++ b/components/feature/qr/src/test/java/mozilla/components/feature/qr/QrFragmentTest.kt
@@ -4,6 +4,7 @@
 
 package mozilla.components.feature.qr
 
+import android.hardware.camera2.CameraAccessException
 import android.hardware.camera2.CameraDevice
 import android.util.Size
 import com.google.zxing.BarcodeFormat
@@ -16,6 +17,7 @@ import mozilla.components.support.test.eq
 import mozilla.components.support.test.mock
 import org.junit.Assert.assertEquals
 import org.junit.Assert.assertNull
+import org.junit.Assert.fail
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.mockito.ArgumentMatchers.anyInt
@@ -130,6 +132,27 @@ class QrFragmentTest {
         camera = mock()
         qrFragment.stateCallback.onError(camera, 0)
         verify(camera).close()
+    }
+
+    @Test
+    fun `catches and handles CameraAccessException`() {
+        val qrFragment = spy(QrFragment.newInstance(mock(QrFragment.OnScanCompleteListener::class.java)))
+
+        var camera: CameraDevice = mock()
+        `when`(camera.createCaptureRequest(anyInt())).thenThrow(CameraAccessException(123))
+        qrFragment.cameraDevice = camera
+
+        val textureView: AutoFitTextureView = mock()
+        `when`(textureView.surfaceTexture).thenReturn(mock())
+        qrFragment.textureView = textureView
+
+        qrFragment.previewSize = mock()
+
+        try {
+            qrFragment.createCameraPreviewSession()
+        } catch (e: CameraAccessException) {
+            fail("CameraAccessException should have been caught and logged, not re-thrown.")
+        }
     }
 
     @Test

